src/python/pants/engine/internals/uuid.py

```diff
@@ -1,28 +1,49 @@
 # Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-import random
 import uuid
-from dataclasses import dataclass, field
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional, cast
 
-from pants.engine.rules import _uncacheable_rule, collect_rules
+from pants.engine.rules import collect_rules, rule
+from pants.util.meta import frozen_after_init
 
 
-@dataclass(frozen=True)
+class UUIDScope(Enum):
+    PER_CALL = "call"
+    PER_SESSION = "session"
```

Comment:
Is there a need for `PER_SESSION` that you anticipate? I imagine to emulate that you could previously give a fixed seed. We weren't using that mechanism, it could arguably go to make this simpler (e.g. no Enum).

Reply:
Yes.

> FYI: This was / will be used to make #10779 as fast as current Pants under Pantsd runs 2+.

UncacheableProcess will accept a scope and we'll use it to cache bootstrap and internal-only interpreter discovery per session to restore performance of #10779.